tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ConfidentialClientApplicationTests.cs

```diff
@@ -399,7 +421,91 @@ await app
                       .WithForceRefresh(true)
                       .ExecuteAsync(CancellationToken.None)
                       .ConfigureAwait(false);
-                Assert.AreNotEqual(cachedAssertion, app.ClientCredential.Assertion);
+                Assert.AreNotEqual(cachedAssertion, app.ClientCredential.CachedAssertion);
+            }
+        }
+
+        [TestMethod]
+        [TestCategory("ConfidentialClientApplicationTests")]
+        public async Task ConfidentialClientUsingClientAssertionClaimsTestAsync()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                httpManager.AddInstanceDiscoveryMockHandler();
+
+                var cert = new X509Certificate2(ResourceHelper.GetTestResourceRelativePath("valid.crtfile"));
+                var app = CreateConfidentialClient(httpManager, cert, 3, CredentialType.CertificateAndClaims);
+
+                var result = await app.AcquireTokenForClient(MsalTestConstants.Scope.ToArray()).ExecuteAsync(CancellationToken.None).ConfigureAwait(false);
+                Assert.IsNotNull(result);
+                Assert.IsNotNull("header.payload.signature", result.AccessToken);
+                Assert.AreEqual(MsalTestConstants.Scope.AsSingleString(), result.Scopes.AsSingleString());
+
+                // make sure user token cache is empty
+                Assert.AreEqual(0, app.UserTokenCacheInternal.Accessor.GetAllAccessTokens().Count());
+                Assert.AreEqual(0, app.UserTokenCacheInternal.Accessor.GetAllRefreshTokens().Count());
+
+                // check app token cache count to be 1
+                Assert.AreEqual(1, app.AppTokenCacheInternal.Accessor.GetAllAccessTokens().Count());
+                Assert.AreEqual(0, app.AppTokenCacheInternal.Accessor.GetAllRefreshTokens().Count()); // no RTs are returned
+
+                // assert client credential
+
+                Assert.IsNotNull(app.ClientCredential.CachedAssertion);
+                Assert.AreNotEqual(0, app.ClientCredential.ValidTo);
+
+                // save client assertion.
+                string cachedAssertion = app.ClientCredential.CachedAssertion;
+                long cacheValidTo = app.ClientCredential.ValidTo;
+
+                result = await app
+                    .AcquireTokenForClient(MsalTestConstants.ScopeForAnotherResource.ToArray())
+                    .ExecuteAsync(CancellationToken.None)
+                    .ConfigureAwait(false);
+
+                Assert.IsNotNull(result);
+                Assert.AreEqual(cacheValidTo, app.ClientCredential.ValidTo);
+                Assert.AreEqual(cachedAssertion, app.ClientCredential.CachedAssertion);
```

Comment:
> Assert.AreEqual(cachedAssertion, app.ClientCredential.CachedAssertion); [](start = 15, length = 72)

imo, this assert is not enough (similar concern for other places). Pls. consider what the right approach is to validate the assertion

Reply:
ok, ill peak inside the assertion and ensure that the provided claims are actually there